File: go/src/pr037/pr037.go
//////////////////////////////////////////////////////////////////////
// Problem 37
//
// Published on Friday, 14th February 2003, 06:00 pm; Solved by 29331
//
// The number 3797 has an interesting property. Being prime itself, it
// is possible to continuously remove digits from left to right, and
// remain prime at each stage: 3797, 797, 97, and 7. Similarly we can
// work from right to left: 3797, 379, 37, and 3.
//
// Find the sum of the only eleven primes that are both truncatable
// from left to right and right to left.
//
// NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
//
//////////////////////////////////////////////////////////////////////
// 748317

package pr037

import "euler"
import "fmt"

func Run() {
	var s euler.Sieve

	count := 0
	sum := 0
	p := 11
	for count < 11 {
		if isRightPrime(&s, p) && isLeftPrime(&s, p) {
			sum += p
			count++
			// fmt.Printf("%d\n", p)
		}

		p = s.NextPrime(p)
	}
	fmt.Printf("%d\n", sum)
}

func isRightPrime(s *euler.Sieve, num int) bool {
	for num > 0 {
		if !s.IsPrime(num) {
			return false
		}
		num /= 10
	}
	return true
}

func isLeftPrime(s *euler.Sieve, num int) bool {
	mod := 1
	for mod < num {
		mod *= 10
	}

	for mod > 1 {
		num %= mod
		mod /= 10
		if !s.IsPrime(num) {
			return false
		}
	}

	return true
}

File: fsharp/pr037.fs
// Problem 37
//
// 14 February 2003
//
//
// The number 3797 has an interesting property. Being prime itself, it is
// possible to continuously remove digits from left to right, and remain
// prime at each stage: 3797, 797, 97, and 7. Similarly we can work from
// right to left: 3797, 379, 37, and 3.
//
// Find the sum of the only eleven primes that are both truncatable from left
// to right and right to left.
//
// NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
//
// 748317

module Pr037

// Solve this by trying and filtering all primes.  This works pretty close to a lazy solution in
// haskell.

// Convenient int-based one.
let isPrime (x : int) = Miller.isPrime (uint64 x) 20

// Generate all of the right removal digits from a given number.
let rec leftGen n =
    if n = 0 then Seq.empty
    else seq {
        yield n
        yield! (leftGen (n / 10))
    }

let rightGen n =
    let rec loop n build ten =
        if n = 0 then Seq.singleton build
        else seq {
            yield build
            yield! (loop (n / 10) ((n % 10) * ten + build) (ten * 10))
        }
    loop (n / 10) (n % 10) 10

// Fortunately the problem statement tells us there are only 11 of these primes, so we can stop
// generating at that point.
let euler037 () =
    let sv = Sieve.Sieve ()
    sv.PrimesFrom 11
    |> Seq.filter (fun x ->
        let lefts = leftGen x
        let rights = rightGen x
        Seq.append lefts rights
        |> Seq.forall (fun n -> isPrime n))
    |> Seq.take 11
    |> Seq.sum

#if COMPILED
euler037 () |> printfn "%A"
#endif
